### PR TITLE
FDG-4227 Spark queue enforcement changes

### PR DESCRIPTION
**Pull request by surajbaghel**
## What changes were proposed in this pull request?

Spark queue enforcement changes

## How was this patch tested?

Unit testing, integration testing and manual testing is done using fk-pf-spark.

### CODE REVIEW

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -227,6 +228,32 @@ private[spark] class Client(
...
+    } catch {
+      case e: Throwable =>
+        throw new RuntimeException("User couldn't be set. ", e)
+    }
+    user
+  }
+
+  @throws[IOException]
+  private def getEnforcedQueue(queue: String): String = {
+    log.info("default queue: " + queue)
```
Comment (kartikaditya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Fix log string
Its input queue name

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 2]
```diff
@@ -227,6 +228,32 @@ private[spark] class Client(
...
+    val queueEnforcerClassName: String = sparkConf.get(QUEUE_ENFORCER_CLASS)
+    log.info("queue enforcer class: " + queueEnforcerClassName)
+    var queueEnforcer: QueueEnforcer = null
+    try {queueEnforcer = Utils.classForName(queueEnforcerClassName)
+      .newInstance.asInstanceOf[QueueEnforcer]
+    } catch {
+      case e: Throwable =>
+        throw new RuntimeException("Couldn't enforce queue. ", e)
+    }
+    queueEnforcer.getEnforcedQueue(queue, getInitiator, sparkConf)
```
Comment (kartikaditya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Can you log the enforced queue

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 3]
```diff
@@ -227,6 +228,32 @@ private[spark] class Client(
...
+        throw new RuntimeException("User couldn't be set. ", e)
+    }
+    user
+  }
+
+  @throws[IOException]
+  private def getEnforcedQueue(queue: String): String = {
+    log.info("default queue: " + queue)
+    val queueEnforcerClassName: String = sparkConf.get(QUEUE_ENFORCER_CLASS)
+    log.info("queue enforcer class: " + queueEnforcerClassName)
```
Comment (kartikaditya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Move this log below and also print enforced queue

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/ReturnExistingQueue.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package org.apache.spark.deploy.yarn.propertymodifier
+
+import org.apache.spark.SparkConf
+
+class ReturnExistingQueue extends QueueEnforcer {
```
Comment (kartikaditya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/ReturnExistingQueue.scala]: Remove this and let's not have it, else user can hack by setting this enforcer and override what ever he/she wants

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/QueueEnforcer.scala [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package org.apache.spark.deploy.yarn.propertymodifier
+
+import org.apache.spark.SparkConf
+
+trait QueueEnforcer {
```
Comment (kartikaditya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/QueueEnforcer.scala]: Lets have only 1 implementation of this which org queue enforcer

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala [hunk 6]
```diff
@@ -73,6 +73,11 @@ package object config {
     .stringConf
     .createWithDefault("default")
 
+  private[spark] val QUEUE_ENFORCER_CLASS = ConfigBuilder("spark.queue.enforcer.class")
+    .doc("Fully Qualified class name for concrete impl to be used to enforce queue")
+    .stringConf
+    .createWithDefault("org.apache.spark.deploy.yarn.propertymodifier.ReturnExistingQueue")
```
Comment (kartikaditya) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala]: Remove this default and fail if its value is not present

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (surajbaghel) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Updated log string.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 3 above)
Comment (surajbaghel) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Moved this log below and also logging enforced queue.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 2 above)
Comment (surajbaghel) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Logging enforced queue.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/ReturnExistingQueue.scala (hunk 4 above)
Comment (surajbaghel) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/ReturnExistingQueue.scala]: Removed

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/QueueEnforcer.scala (hunk 5 above)
Comment (surajbaghel) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/propertymodifier/QueueEnforcer.scala]: Now having only 1 implementation which is com.flipkart.fdp.orgqueue.OrgQueueEnforcerForInitiator

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala (hunk 6 above)
Comment (surajbaghel) [reply resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala]: Done changes.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 7]
```diff
@@ -227,6 +228,39 @@ private[spark] class Client(
...
+    }
+
+    var queueEnforcer: QueueEnforcer = null
+    try {queueEnforcer = Utils.classForName(queueEnforcerClassName)
+      .newInstance.asInstanceOf[QueueEnforcer]
+    } catch {
+      case e: Throwable =>
+        throw new RuntimeException("Couldn't enforce queue. ", e)
+    }
+    log.info("queue enforcer class: " + queueEnforcerClassName)
```
Comment (mittalnanu) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: change the log to debug here./

**Pull request merged**